### DISCUSSION TITLE
Theme Swapping

### DISCUSSION THREAD

**Discussion by BlueMystical** [BlueMystical/Dark-Mode-Forms#38] [Show and tell]
Hi there.
I just made some progress on the requested feature for on the fly Theme Swapping

On this Example:

1. Theme detected automatically, Dark Mode On at Windows Settings.
2. Swap theme button clicked.   Dark to Clear Mode.
3. Swap theme button clicked.   Clear to Dark Mode.

![1](https://github.com/user-attachments/assets/aaf19fe4-443e-4a76-96bb-ae82665ca744)

Ther are still some colors that are refusing to change, gotta find them, but it works.
Will commit this progress in a few.

Regards.

**Comment by jukfiuune** [comment]
Very nice! Question - How exactly do you change the color of the scroll bars to the dark one? i remember just setting immersivedarkmode to 1 just changed the titlebar style, not the scrollbars. Thanks

**Reply by BlueMystical** [reply]
i use a combination of the `SetWindowTheme` and `DwmSetWindowAttribute` Win32 APIs, you can check the actual [code here](https://github.com/BlueMystical/Dark-Mode-Forms/blob/main/SourceFiles/DarkModeCS.cs#L900).

**Comment by BlueMystical** [comment]
![image](https://github.com/user-attachments/assets/3f5499c2-b09b-4197-9359-9c905446e641)
Good news!

On the fly theme swapping is now fully enabled, see the Example project.
Also the ComboBox Control is now fully themed, the 'FlatComboBox' is no longer needed.

Regards.

**Reply by MagicAndre1981** [reply]
for me it doesn't work. Do I have to enable a setting to react to theme change in windows? or do you mean the user can swap the theme on its own and this works on the fly?

**Reply by MagicAndre1981** [reply]
ok, I can call `ApplyTheme(true);` to force dark mode no matter which theme is selected so my request from #30 is implemeted. Great 🥳

**Reply by BlueMystical** [reply]
Yep, the `ApplyTheme` method is the way, check the '`Theme Swap`' button on the Example Form.

**Reply by MagicAndre1981** [reply]
now that I can force a theme can we also handle theme change of windows settings app? I tried to override WndProc and handle WM_THEMECHANGED but the code is never executed. Is changing dark/light mode raising a different message?

**Comment by kachnitata** [comment]
hi, one of the recent commits seems to have a little side effect.

with using

`			switch (control.GetType().Name)
`

instead of former `is` operator, now the inherited controls do not work as expected.

I am using a Control derived from `TabControl` that allows to close individual tabs and now it does not look nice:


![before](https://github.com/user-attachments/assets/41868a26-7d48-4787-82d6-8c86b72c8e57)

![after](https://github.com/user-attachments/assets/a447de76-67cb-4dde-9ef6-3ed7b562411c)

are there any plans to handle such cases?

**Reply by BlueMystical** [reply]
i considered that a Switch Cases was better than a pile of Ifs, i also were having a weird issue with label controls who for some reason were misbehaving with the IFs, the Switch solved those issues and provide better code reading.

As for the Inherited tab control, it is weird, because the `FlatTabControl` (on the right side) i use is also inheriterited and its not having this issue:
![image](https://github.com/user-attachments/assets/8ff5b769-43df-4e43-9f11-829d25b1e007)

Im gonna do some more tests to try to replicate your situation and if i do, i can find a solution.

Regards.

**Reply by kachnitata** [reply]
the problem is that my `TabControlEx` uses custom `OnDrawItem` override in order to paint the "X" for closing the tab.
When I started using DarkModeCS I spent quite a few hours with accomodating it..

I have nothing against Switch, I think it is a bit clearer, but I am wondering how to adjust it for the inherited controls.
One option would be, instead of `control.GetType().Name` using something like

```
      Type t0 = control.GetType();
      Type t = t0;
      //try to find the closest ancestor belonging to System.Windows.Forms or DarkModeForms
      while (t != typeof(Object) && !t.Namespace.StartsWith("System.Windows.Forms") && !t.Namespace.Equals("DarkModeForms")) t = t.BaseType;
      //if not find, stay with the current type
      if(t == typeof(Object)) t = t0;

      switch (t.Name)

```

I have briefly tested it and it seems to work for me. The last "if branch" is not actually executed in my app.

**Reply by BlueMystical** [reply]
wait, i may had found it, on this version i added this line:
![image](https://github.com/user-attachments/assets/f4327ddc-8ec1-4114-9ad7-4381a4eed37e)
try removing it.

**Reply by kachnitata** [reply]
unfortunately this does not help. On the other hand, the code from my previous comment works for me

**Reply by kachnitata** [reply]
I think the problem really is with the GetType approach - so that the inherited types are not recognized and handled properly, that's why I suggested traversing the inheritance hierarchy until we find a known type (provided by Microsoft or DarkModeForms)

**Reply by kachnitata** [reply]
also, I just found that the current version (even with my abovementioned tweak) does not handle the multi-level context menus again, only the first level is themed: 
I will try to look at it.

![image](https://github.com/user-attachments/assets/de80b5f7-45d8-4549-b658-913e5c344d45)

**Reply by BlueMystical** [reply]
alright, thanks for the feedback
I'll see how can i fix this.
Regards.

**Reply by kachnitata** [reply]
I found it. the problem is `ContextMenuStrip` is a subclass of `ToolStripDropDownMenu` which is a subclass of `ToolStripDropDown` .

in the older version, both `ContextMenuStrip` and `ToolStripDropDown` sections were executed for a `ContextMenuStrip`.
Now, thanks to the `switch` approach, only the first one.

**Reply by kachnitata** [reply]
which leads me to 

```
				case "ToolStripDropDown":
					(control as ToolStripDropDown).Opening += Tsdd_Opening;
					break;
				case "ContextMenuStrip":
					(control as ContextMenuStrip).RenderMode = ToolStripRenderMode.Professional;
					(control as ContextMenuStrip).Renderer = new MyRenderer(new CustomColorTable(OScolors), ColorizeIcons) { MyColors = OScolors };
          (control as ContextMenuStrip).Opening += Tsdd_Opening;
					break;

```

this solved it partly. I will try to fix completely tomorrow.

**Reply by BlueMystical** [reply]
Nice find!!

**Reply by kachnitata** [reply]
the previous fix dealt with the first two levels only; the completely working version is:
```

        case "ToolStripDropDownMenu":
          (control as ToolStripDropDownMenu).Opening += Tsdd_Opening;
          break;
				case "ContextMenuStrip":
					(control as ContextMenuStrip).RenderMode = ToolStripRenderMode.Professional;
					(control as ContextMenuStrip).Renderer = new MyRenderer(new CustomColorTable(OScolors), ColorizeIcons) { MyColors = OScolors };
          (control as ContextMenuStrip).Opening += Tsdd_Opening;
					break;

```

as you can see, I replaced` case "ToolStripDropDownMenu":` which was not actually executed with `case "ToolStripDropDownMenu":`. Now the menus looks perfect for arbitrary number of levels.


should I create PR for my suggested changes or would you consider including them yourself?

**Reply by BlueMystical** [reply]
Its done, thank you for all your contributions. 
Regards.

**Reply by kachnitata** [reply]
thanks @BlueMystical,
unfortunately, this leaves both the problems unsolved, this is just a fraction of the necessary changes.
I have again adjusted the file and I will post a PR, so take a look, pls
regards

**Reply by kachnitata** [reply]
I made a mistake when pulling the current version. So context menus are ok now, but for the inherited control issue I suggest using

`      switch (GetKnownUIType(control.GetType()).Name)
`
and 

```
    /// <summary>
    /// in order to be able to use Type.Name we have to find a "known" type
    /// this handles the cases of custom controls derived from those provided by WinForms or DarkModeForms
    /// so we try to find the closest ancestor belonging to System.Windows.Forms or DarkModeForms
    /// if not found, we just return the current type
    /// </summary>
    private static Type GetKnownUIType(Type t0)
    {
      Type t = t0;
      while (t != typeof(Object) && !t.Namespace.StartsWith("System.Windows.Forms") && !t.Namespace.Equals("DarkModeForms")) t = t.BaseType;
      return t == typeof(Object) ? t0 : t;
    }

```

please consider it, or let me know if you suggest other solution
regards

**Reply by kachnitata** [reply]
possible alternative would be some kind of "event handler" for extending the functionality for unhandled types, or dictionary which would say "treat type X as Y" or something like that....

**Reply by kachnitata** [reply]
Actually, I start liking the idea... The user would call a Register method which would record into a Dictionary the type name pair. It would be looked up before (from) the Case statement. Would be safer and more transparent than my current solution. Should I elaborate and post the code?

**Reply by kachnitata** [reply]
In case the user does not Register anything, the dictionary is empty and everything works unmodified

**Reply by BlueMystical** [reply]
This issue seems to be escalating too high, maybe i just revert to the pile of IFs...

**Comment by BlueMystical** [comment]
Alright, v1.8.3 reverts to the old trusty pile of IFs
Hope this fixes the issues
Regards.
![Sin título](https://github.com/user-attachments/assets/47c08c2a-aba5-4d17-a025-8a9d4052ef21)

**Reply by BlueMystical** [reply]
and sorry for the late, i had a lot of RL work lately.

**Comment by BlueMystical** [comment]
By the way @kachnitata just added an optional Close Tab button on the FlatTab Control, check it out !

![image](https://github.com/user-attachments/assets/e38c6688-f4ee-4e31-a721-31d03c418e12)

Feel free to modify it as you need in your own project

**Reply by kachnitata** [reply]
Wow, thank you !
